cmd/event-received/main.go

```diff
@@ -155,63 +177,118 @@ func handleFeeApproved(ctx context.Context, dynamoClient dynamodbClient, event e
-		return fmt.Errorf("failed to unmarshal 'more-evidence-required' detail: %w", err)
+		return fmt.Errorf("failed to unmarshal '%s' detail: %w", moreEvidenceRequiredEventName, err)
 	}
 
-	var key dynamo.Key
-	if err := client.OneByUID(ctx, v.UID, &key); err != nil {
-		return fmt.Errorf("failed to resolve uid for 'more-evidence-required': %w", err)
+	lpa, err := getLpaByUID(ctx, client, v.UID, moreEvidenceRequiredEventName)
+	if err != nil {
+		return err
 	}
 
-	if key.PK == "" {
-		return errors.New("PK missing from LPA in response to 'more-evidence-required'")
+	lpa.Tasks.PayForLpa = actor.PaymentTaskMoreEvidenceRequired
+	lpa.UpdatedAt = now()
+
+	if err := client.Put(ctx, lpa); err != nil {
+		return fmt.Errorf("failed to update LPA task status for '%s': %w", moreEvidenceRequiredEventName, err)
 	}
 
-	var lpa page.Lpa
-	if err := client.One(ctx, key.PK, key.SK, &lpa); err != nil {
-		return fmt.Errorf("failed to get LPA for 'more-evidence-required': %w", err)
+	return nil
+}
+
+func handleFeeDenied(ctx context.Context, client dynamodbClient, event events.CloudWatchEvent, now func() time.Time) error {
+	var v uidEvent
+	if err := json.Unmarshal(event.Detail, &v); err != nil {
+		return fmt.Errorf("failed to unmarshal '%s' detail: %w", feeDeniedEventName, err)
 	}
 
-	lpa.Tasks.PayForLpa = actor.PaymentTaskMoreEvidenceRequired
+	lpa, err := getLpaByUID(ctx, client, v.UID, feeDeniedEventName)
+	if err != nil {
+		return err
+	}
+
+	lpa.Tasks.PayForLpa = actor.PaymentTaskDenied
 	lpa.UpdatedAt = now()
 
 	if err := client.Put(ctx, lpa); err != nil {
-		return fmt.Errorf("failed to update LPA task status for 'more-evidence-required': %w", err)
+		return fmt.Errorf("failed to update LPA task status for '%s': %w", feeDeniedEventName, err)
 	}
 
 	return nil
 }
 
-func handleFeeDenied(ctx context.Context, client dynamodbClient, event events.CloudWatchEvent, now func() time.Time) error {
-	var v uidEvent
+func handleObjectTagsAdded(ctx context.Context, client dynamodbClient, event events.CloudWatchEvent, now func() time.Time, s3Client s3Client) error {
+	var v objectTagsAddedEvent
 	if err := json.Unmarshal(event.Detail, &v); err != nil {
-		return fmt.Errorf("failed to unmarshal 'fee-denied' detail: %w", err)
+		return fmt.Errorf("failed to unmarshal '%s' detail: %w", objectTagsAddedEventName, err)
 	}
 
-	var key dynamo.Key
-	if err := client.OneByUID(ctx, v.UID, &key); err != nil {
-		return fmt.Errorf("failed to resolve uid for 'fee-denied': %w", err)
+	objectKey := v.Object.Key
+
+	tags, err := s3Client.GetObjectTags(ctx, objectKey)
+	if err != nil {
+		return fmt.Errorf("failed to get tags for object in '%s': %w", objectTagsAddedEventName, err)
 	}
 
-	if key.PK == "" {
-		return errors.New("PK missing from LPA in response to 'fee-denied'")
+	hasScannedTag := false
+	var tagIndex int
+
+	for i, tag := range tags {
+		if *tag.Key == "virus-scan-status" {
+			hasScannedTag = true
+			tagIndex = i
```

Comment:
instead of storing the index to get later, would it be easier to do something like

```go
hasScannedTag := false
hasVirus := false
for _, tag := range tags {
  if *tag.Key == "virus-scan-status" {
    hasScannedTag = true
    hasVirus = tag.Value == virusFound
    break
  }
}
```